admin/app/com/lucidchart/piezo/admin/utils/CronHelper.scala

```diff
@@ -0,0 +1,185 @@
+      // subtract an hour to avoid double counting innerMaxIntervalSeconds
+      sampledMaxIntervalInSeconds - HourUnitType.secondsPerUnit + innerMaxIntervalSeconds
+    }
+  }
+
+  /**
+   * Estimates the max interval for a given cron expression. The more samples, the better the estimate.
+   */
+  private def getSampledMaxInterval(prev: Date, numSamples: Long, trigger: CronTrigger, maxInterval: Long = 0): Long = {
+    Option(trigger.getFireTimeAfter(prev)) match {
+      case Some(next) if numSamples > 0 =>
+        val intervalInSeconds = next.toInstant.getEpochSecond - prev.toInstant.getEpochSecond
+        if (intervalInSeconds > maxInterval) {
+          val sampleId = NUM_COMPLEX_SAMPLES - numSamples
+          logger.debug(s"Seconds:$intervalInSeconds Sample:$sampleId Interval:$prev -> $next")
+        }
+        getSampledMaxInterval(next, numSamples - 1, trigger, Math.max(intervalInSeconds, maxInterval))
+      case _ => maxInterval
+    }
+  }
+
+  private def getCronContainers(parts: Array[String], timeZone: Option[TimeZone]): List[CronContainer] = {
+    parts
+      .zip(List(SecondUnitType, MinuteUnitType, HourUnitType))
+      .map { case (str, cronType) => CronContainer(str, cronType, timeZone) }
+      .toList
+  }
+
+}
+
+sealed abstract class UnitType(val secondsPerUnit: Long, val numUnitsInContainer: Long)
+case object SecondUnitType extends UnitType(1, 60)
+case object MinuteUnitType extends UnitType(60, 60)
+case object HourUnitType extends UnitType(3600, 24)
+
+/**
+ * A cron container is composed of "units". The "unitType" determines how many units are in the container, and how many
+ * seconds are in each unit. "Marked units" describe when the trigger is set to fire. "Intervals" describe the number of
+ * units between (but not including) marked units.
+ */
+case class CronContainer(str: String, unitType: UnitType, markedUnits: List[Long], timeZone: TimeZone) {
+  lazy val areAllUnitsMarked: Boolean = markedUnits.size == unitType.numUnitsInContainer
+
+  // the "wrap interval" is the interval that wraps around both ends of the container
+  private lazy val wrapInterval = (unitType.numUnitsInContainer - markedUnits.last) + markedUnits.head
+  lazy val wrapIntervalInSeconds: Long = unitsToSeconds(wrapInterval)
+
+  // the max interval is the longest interval between any two marked units in the container, including the wrap interval
+  lazy val maxIntervalInSeconds: Long = {
+    val otherIntervals = markedUnits.zipWithIndex
+      .take(markedUnits.size - 1)
+      .map { case (value, index) => markedUnits(index + 1) - value }
+    val units = (otherIntervals :+ wrapInterval).max
+    val unitsWithDaylightSavings = units + (if (unitType == HourUnitType) getDaylightSavings(units, markedUnits) else 0)
+    unitsToSeconds(unitsWithDaylightSavings)
+  }
+
+  /**
+   * We multiply the number of units by the number of seconds in a single unit. We also subtract 1 unit to avoid double
```

Comment:
Actually I  think what we can do here is expose a value that is the time between the last trigger time and the first trigger time (markedUnits.last - markedUnits.first), and then just implement `unitsToSeconds` as expected  here. Then instead of adding `wrappingInterval` we subtract this new value (totalInterval?) That is mathematically equivalent to what you currently have.

Reply:
That's pretty clever - great idea!